### PR TITLE
Add snprintf and remove old sprintf

### LINKED ISSUE #3287: Some numbers do not roundtrip through sjson correctly.
### Expected behavior

If `a` is a lua number then
```
a == sjson.decode(sjson.encode({a}))[1]
```

Specific example:
```
a = -4392806464851163000000000

rt_a = sjson.decode(sjson.encode({a}))[1]

if a ~= rt_a then
   print ("Different", a, rt_a, a-rt_a)
end
```

### Actual behavior

The code above prints:
```
Different	-4.392806e+24	-4.392806e+24	-5.764608e+17
```

### NodeMCU startup banner
```
NodeMCU 3.0.0.0 built with Docker provided by frightanic.com
	branch: float-test
	commit: 9d74cd5aa0960c10a9d8b04196d049023281e908
	release: 3.0-master_20200910 +4
	release DTS: 202009182147
	SSL: false
	build type: float
	LFS: 0x0 bytes total capacity
	modules: adc,bit,dht,enduser_setup,file,gpio,i2c,mqtt,net,node,ow,sjson,spi,tmr,uart,wifi
 build 2020-09-20 13:08 powered by Lua 5.3.5 on SDK 3.0.1-dev(fce080e)
```

I had to implement a couple of extra methods to be able to create uniformly distributed floating point values. I may well do a PR for those (simple) functions (which are under `#ifdef DEVELOPMENT_TOOLS`)

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #3287.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.

This replaces the in-tree implementation of sprintf with snprintf (from https://github.com/weiss/c99-snprintf ). This is added as a submodule.

There are still places that call sprintf that ought to be calling snprintf.

### CODE REVIEW

**Comment by pjsg** [comment]
Additionally, it won't format (in f format) numbers bigger than 10^20. This isn't really a problem (I think).

**Comment by vsky279** [comment]
I tested this PR with tests that were discussed in recent PRs concerning `stdio.c`

```Lua
for i=-1,1 do print(("%.3f\t%+.3f\t% .3f"):format(0.0004*i, 0.0004*i, 0.0004*i)) end
for i=1,20 do d=(3.1405+i*0.001); print(("%.3f\t%7f\t%.20f"):format(d, d, d)) end
for i=1,10 do d=-(3.1405+i*0.001)*1e-5; print(("%.3e\t%7e\t%.20e"):format(d, d, d)) end
```
All prints as expected.

Also the sjson roundtrip test passes well even for numbers where it was failing with previous version
```Lua
a={}
for i=1,100 do a[i]=math.floor(node.random()*1e21)/(10^(math.floor(math.random()*20)-10)) end
b=sjson.encode(a)
c=sjson.decode(b)
for i,j in pairs(a) do if a[i]~=c[i] then print(("mismatch [%d]: %.20g ~= %.20g, diff: %.20g"):format(i, a[i], c[i], a[i]-c[i])) end end
```

`%d`, `%s` formats seem to work fine too.

I'd like to implement double rounding `#define SPRINTF_ROUNDING_MODE_DOUBLEROUNDING` as discussed in #3283 once this lands in `dev`.

File: app/lua53/luaconf.h
Comment (nwf) [inline app/lua53/luaconf.h:404]: Is `LUA_USE_C89` not defined when we're building for the ESP8266?

File: app/modules/node.c
Comment (nwf) [inline app/modules/node.c:647]: The changes to this file are not really part of this PR (at least, not by title), but I see no reason they shouldn't also land.

File: app/libc/snprintf.c
Comment (nwf) [inline app/libc/snprintf.c:1571]: Can the test code move to its own file?

**Review by nwf — CHANGES_REQUESTED** [review]
Mostly LGTM.

Any possibility we could rip `libmain.o`'s `eagle_lib.o` out of our builds if we're carrying around our own `v?snprintf`?  Annoyingly, there are some other symbols (`ets_strchr`, `ets_strcat`, `bit_popcount`???) in there that we'd need to add elsewhere.  Espressif continues to disappoint.

File: app/lua53/luaconf.h
Comment (pjsg) [reply app/lua53/luaconf.h:404]: I'll fix that.

File: app/modules/node.c
Comment (pjsg) [reply app/modules/node.c:647]: I added these as part of testing the new behavior. I can certainly rip them out (I didn't actually intend to commit them). I didn't add documentation for them. I think that they are only relevant for lua53 as well....

What would you like me to do?

File: app/libc/snprintf.c
Comment (pjsg) [reply app/libc/snprintf.c:1571]: It could do -- but this file is essentially the exact code from the c99-snprintf repo -- so I'd rather leave it that way.

**Comment by pjsg** [comment]
I'm try to figure out how to exclude eagle_lib.o from the linker but not
having any luck at the moment....

File: app/modules/node.c
Comment (nwf) [reply app/modules/node.c:648]: I think it's fine to leave them in this PR; I was just noting.

File: app/libc/snprintf.c
Comment (nwf) [reply app/libc/snprintf.c:1571]: If that's so, could we make this a `git submodule` instead?

**Comment by nwf** [comment]
I appreciate you trying; let's not hold up this PR behind my crazed desire to divorce ourselves from Espressif. :)

**Review by nwf — APPROVED** [review]

**Comment by nwf** [comment]
With the `git submodule` landed and CI happy, is this ready to land?

**Comment by pjsg** [comment]
I think so. I have a followup PR to try and get rid of most/all of the
sprintf usages -- that will require more testing as some of the length
calculations are complex.

**Pull request merged**
